crypto: misc refactorings

bnjbvr opened the pull request:
This started as a partial attempt to fix #1415, but the commit that was supposed to fix this has been removed from this PR, as it wasn't deemed useful: only refactorings remain, each explicited as an individual commit.

For history, the previous text for that PR stands below:

---

We're using application-level transactions to make sure that the account is properly synchronized in the cache vs in the database.

Before this commit, the transaction would be committed only when *all* the operations in it succeeded. This was based on the assumption that most encryption requests could be replayed, by re-sending them to the server. Unfortunately, this assumption doesn't hold for when generating one-time keys: it could be that one time-keys would be generated by the client, then the application-level transaction would fail, resulting in the client "forgetting" about the one time keys it uploaded. The server rejects reuploads of existing one-time keys, so that would end up wedging a device, causing unable-to-decrypt events, without a proper way out.

Here, we propose to save the account just after one-time keys have been generated, in a separate transaction.

A partial attempt to address #1415.

cc @kegsay @richvdh @BillCarsonFr

richvdh commented:
> The server rejects reuploads of existing one-time keys

No. The server rejects uploads of a *different* one-time-key with the same ID as an existing one.

> , so that would end up wedging a device, causing unable-to-decrypt events

The fact that the server is catching the error is actually *saving* you from UTDs. The real problem with uploading OTKs before they are persisted to the db is this scenario:

 * Alice's client generates an OTK but does not yet persist it to the db.
 * Alice's client uploads the OTK to the server.
 * Alice's client is killed (OOM? device is turned off?) before the key is persisted, losing all record of that OTK
 * Bob attempts to send a message to Alice, so claims the OTK
 * Alice, having restarted her client, receives the message from Bob. But it is encrypted with an OTK she has no record of. UTD error follows.

richvdh commented on crates/matrix-sdk-crypto/src/machine.rs:
```diff
@@ -1215,23 +1213,27 @@ impl OlmMachine {
+        // Updating key counts may generate new one-time keys, and we should *never*
+        // forget about them once generated (otherwise this leads to issues like
+        // #1415), so save them in a separate transaction here.
+        self.inner
+            .store
+            .with_transaction(|mut transaction| async {
+                let account = transaction.account().await.map_err(OlmError::Store)?;
+                account.update_key_counts(
+                    sync_changes.one_time_keys_counts,
+                    sync_changes.unused_fallback_keys,
+                );
+                Ok((transaction, ()))
+            })
+            .await?;
```
I'm afraid I'm failing to follow how this fixes the problem.

Is there not still a possibility that committing the transaction can fail, leaving us in the same situation as before? (As I understand it, that situation is that the  `Account` has been updated, but not persisted to the database.)

Hywan replied:
Hmm that's correct that this transaction can fail, if the account cannot be fetched. The rest of the code cannot fail.

Hywan replied:
One solution would be to fetch the account before running the transaction maybe (if possible, because for now, the account is fetched by using the transaction).

bnjbvr replied:
We haven't quite identified the cause of the problem yet, so we're still navigating a bit randomly here, and two days ago we *thought* this could be the source, but it's probably not.

But the *only* way to modify an `Account` is to get a transaction, that gives us a write-access scope to the account. There's no other way to upsert `Account`s in theory; in practice, that needs to be double-checked because I'm not sure the API to write `Account` to the database is private/hidden/limited.

richvdh reviewed (changes requested):
sorry, maybe I'm being stupid

richvdh commented:
[Could this also fix https://github.com/matrix-org/matrix-rust-sdk/issues/2998? If not, it's not a complete fix to #1415. Have updated the description on the assumption it is not.]

bnjbvr commented:
For what it's worth, I'm retargeting this PR to only include the drive-by refactorings I've added, and I've removed the commit that was attempting to fix, because it's pretty certain that it's not a fix.

As long as we don't have a clear repro for this issue, it seems impossible to fix it.
